src/pe/exception.rs

```diff
@@ -0,0 +1,810 @@
+        self.len() == 0
+    }
+
+    /// Iterates all function entries in order of their code offset.
+    ///
+    /// To search for a function by relative instruction address, use [`find_function`]. To resolve
+    /// unwind information, use [`get_unwind_info`].
+    ///
+    /// [`find_function`]: struct.ExceptionData.html#method.find_function
+    /// [`get_unwind_info`]: struct.ExceptionData.html#method.get_unwind_info
+    pub fn functions(&self) -> RuntimeFunctionIterator<'a> {
+        RuntimeFunctionIterator {
+            data: &self.bytes[self.offset..self.offset + self.size],
+        }
+    }
+
+    /// Returns the function at the given index.
+    pub fn get_function(&self, index: usize) -> error::Result<RuntimeFunction> {
+        self.get_function_by_offset(index * RUNTIME_FUNCTION_SIZE)
+    }
+
+    /// Performs a binary search to find a function entry covering the given RVA relative to the
+    /// image.
+    pub fn find_function(&self, rva: u32) -> error::Result<Option<RuntimeFunction>> {
+        // NB: Binary search implementation copied from std::slice::binary_search_by and adapted.
+        // Theoretically, there should be nothing that causes parsing runtime functions to fail and
+        // all access to the bytes buffer is guaranteed to be in range. However, since all other
+        // functions also return Results, this is much more ergonomic here.
+
+        let mut size = self.len();
+        if size == 0 {
+            return Ok(None);
+        }
+
+        let mut base = 0;
+        while size > 1 {
+            let half = size / 2;
+            let mid = base + half;
+            let offset = self.offset + mid * RUNTIME_FUNCTION_SIZE;
+            let addr = self.bytes.pread_with::<u32>(offset, scroll::LE)?;
+            base = if addr > rva { base } else { mid };
+            size -= half;
+        }
+
+        let offset = self.offset + base * RUNTIME_FUNCTION_SIZE;
+        let addr = self.bytes.pread_with::<u32>(offset, scroll::LE)?;
+        let function = match addr.cmp(&rva) {
+            Ordering::Less | Ordering::Equal => self.get_function(base)?,
+            Ordering::Greater if base == 0 => return Ok(None),
+            Ordering::Greater => self.get_function(base - 1)?,
+        };
+
+        if function.end_address > rva {
+            Ok(Some(function))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Resolves unwind information for the given function entry.
+    pub fn get_unwind_info(
+        &self,
+        mut function: RuntimeFunction,
+        sections: &[section_table::SectionTable],
+    ) -> error::Result<UnwindInfo> {
+        while function.unwind_info_address % 2 != 0 {
+            let rva = (function.unwind_info_address & !1) as usize;
+            function = self.get_function_at(rva, sections)?;
+        }
+
+        let rva = function.unwind_info_address as usize;
+        let offset = utils::find_offset(rva, sections, self.file_alignment).ok_or_else(|| {
```

Comment:
Can we store the RVA bias in `ExceptionData` instead of looking it up every time?

Reply:
Excellent point. I wasn't sure if the RVAs always point into the same section. This would make sense generally, but I don't trust this is the case for every PE. Especially the tooling Microsoft applies to their DLLs might invalidate this assumption. I'll try to investigate this.

Reply:
Yeah I think you're right. breakpad doesn't appear to do anything clever either.